src/renderer/api/kube-watch-api.ts

```diff
@@ -1,201 +1,326 @@
-      console.log("%cKUBE-WATCH-API:", `font-weight: bold`, ...data);
-    }
+    return "";
   }
 
-  addListener(store: KubeObjectStore, callback: (evt: IKubeWatchEvent) => void) {
-    const listener = (evt: IKubeWatchEvent<KubeJsonApiData>) => {
-      if (evt.type === "ERROR") {
-        return; // e.g. evt.object.message == "too old resource version"
-      }
+  protected getMessage(event: IKubeWatchEvent): IKubeWatchMessage {
+    const message: IKubeWatchMessage = {};
+
+    switch (event.type) {
+      case "ADDED":
+      case "DELETED":
 
-      const { namespace, resourceVersion } = evt.object.metadata;
-      const api = apiManager.getApiByKind(evt.object.kind, evt.object.apiVersion);
+      case "MODIFIED": {
+        const data = event as IKubeWatchEvent<KubeJsonApiData>;
+        const api = apiManager.getApiByKind(data.object.kind, data.object.apiVersion);
 
-      api.setResourceVersion(namespace, resourceVersion);
-      api.setResourceVersion("", resourceVersion);
+        message.data = data;
 
-      ensureObjectSelfLink(api, evt.object);
+        if (api) {
+          ensureObjectSelfLink(api, data.object);
 
-      if (store == apiManager.getStore(api)) {
-        callback(evt);
+          const { namespace, resourceVersion } = data.object.metadata;
+
+          api.setResourceVersion(namespace, resourceVersion);
+          api.setResourceVersion("", resourceVersion);
+
+          message.api = api;
+          message.store = apiManager.getStore(api);
+        }
+        break;
       }
-    };
 
-    this.onData.addListener(listener);
+      case "ERROR":
+        message.error = event as IKubeWatchEvent<KubeJsonApiError>;
+        break;
 
-    return () => this.onData.removeListener(listener);
+      case "STREAM_END": {
+        this.onServerStreamEnd(event as IKubeWatchEventStreamEnd);
+        break;
+      }
+    }
+
+    return message;
   }
 
-  reset() {
-    this.subscribers.clear();
+  protected async onServerStreamEnd(event: IKubeWatchEventStreamEnd) {
+    const { apiBase, namespace } = parseKubeApi(event.url);
+    const api = apiManager.getApi(apiBase);
+
+    if (api) {
+      try {
+        await api.refreshResourceVersion({ namespace });
+        this.connect();
+      } catch (error) {
+        this.log({
+          message: new Error(`Failed to connect on single stream end: ${error}`),
+          meta: { event, error },
+        });
+
+        if (this.isActive) {
+          setTimeout(() => this.onServerStreamEnd(event), 1000);
+        }
+      }
+    }
+  }
+
+  protected log({ message, meta = {} }: IKubeWatchLog) {
+    if (isProduction) {
+      return;
+    }
+
+    const logMessage = `%c[KUBE-WATCH-API]: ${String(message).toUpperCase()}`;
+    const isError = message instanceof Error;
+    const textStyle = `font-weight: bold; ${isError ? "color: red;" : ""}`;
+    const time = new Date().toLocaleString();
+
+    if (isError) {
+      console.error(logMessage, textStyle, { time, ...meta });
```

Comment:
Aren't renderer errors already printed with colour red?

Reply:
Yes, good point. Seems redundant.

Reply:
Is there a reason why we also want it to be bold?

Reply:
Oops, answered in other thread.

Reply:
What do you mean by common style?

Reply:
css-style for `console.log`

Reply:
No I guess what I mean is, why do we want them to be styled at all, shouldn't the prefix be good enough?

Reply:
This is done for better visibility in chrome's `devtools` console. Why you think to simplify it? Info about watch-api not so important to notify for debugging?

Reply:
in devtools, `console.error` already shows up as "red" is that not visible enough?

Reply:
Yes, but it's same for all kinds of error in output. 
Why you against bold-text in error? Red color already removed from text-style.

Reply:
I guess what I am really asking about is why having the function at all, yes `logger` shouldn't be in `main` because it gets used in `renderer` too but what it wrong with just calling that?

We already have a decently common style of having a static for the logging prefix, adding metadata already works.

Reply:
`src/main/logger` doesn't work well in `rendrerer` or it might need some tweaks / additional transports..

Reply:
How does it not work? We have been (ab)using it for several releases. Namely there are currently 13 files in renderer who import `main/logger`

Reply:
Anyway wouldn't the better solution just to make a new winston logger instead of hand rolling our own?

Reply:
Yes, will be fixed in follow-up PR(s).

Reply:
Tried to replace to `common/logger` but have not satisfied with the result:
This is funny, but there is no official support for browsers for `winston` since 2013 ([issue](https://github.com/winstonjs/winston/issues/287) and customizing `transport` is a bit tricky (I don't want to install extra npm package for that).

**Before:**
<img width="563" alt="Screenshot 2021-02-01 at 17 25 13" src="https://user-images.githubusercontent.com/6377066/106479029-8730a300-64b2-11eb-97dd-80749cd6d194.png">

**After:**
<img width="562" alt="Screenshot 2021-02-01 at 17 23 39" src="https://user-images.githubusercontent.com/6377066/106479024-85ff7600-64b2-11eb-9bb7-674a248ac1f9.png">